buidl/shamir.py

```diff
@@ -0,0 +1,329 @@
+            cur = (cur << 1) ^ cur
+            if cur > 255:
+                cur ^= 0x11B
+
+    def __init__(self, shares):
+        self.shares = shares
+        if len(shares) > 1:
+            # check that the identifiers are the same
+            ids = {s.id for s in shares}
+            if len(ids) != 1:
+                raise TypeError("Shares are from different secrets")
+            # check that the exponents are the same
+            exponents = {s.exponent for s in shares}
+            if len(exponents) != 1:
+                raise TypeError("Shares should have the same exponent")
+            # check that the k-of-n is the same
+            k = {s.group_threshold for s in shares}
+            if len(k) != 1:
+                raise ValueError("K of K-of-N should be the same")
+            n = {s.group_count for s in shares}
+            if len(n) != 1:
+                raise ValueError("N of K-of-N should be the same")
+            # check that the share lengths are the same
+            lengths = {s.share_bit_length for s in shares}
+            if len(lengths) != 1:
+                raise ValueError("all shares should have the same length")
+            # check that the x coordinates unique
+            xs = {s.group_index for s in shares}
+            if len(xs) != len(shares):
+                raise ValueError("Share indices should be unique")
+        self.id = shares[0].id
+        self.salt = b"shamir" + int_to_big_endian(self.id, 2)
+        self.exponent = shares[0].exponent
+        self.group_threshold = shares[0].group_threshold
+        self.group_count = shares[0].group_count
+        self.share_bit_length = shares[0].share_bit_length
+
+    def decrypt(self, encrypted, passphrase=b""):
+        if len(encrypted) % 2:
+            raise ValueError("encrypted data should be an even number of bytes")
+        else:
+            half = len(encrypted) // 2
+        left = encrypted[:half]
+        right = encrypted[half:]
+        for i in (b"\x03", b"\x02", b"\x01", b"\x00"):
+            f = pbkdf2_hmac(
+                "sha256",
+                i + passphrase,
+                self.salt + right,
+                2500 << self.exponent,  # 10000 << self.exponent // 4
+                dklen=half,
+            )
+            left, right = right, bytes(x ^ y for x, y in zip(left, f))
+        return right + left
+
+    @classmethod
+    def encrypt(cls, payload, id, exponent, passphrase=b""):
+        if len(payload) % 2:
+            raise ValueError("payload should be an even number of bytes")
+        else:
+            half = len(payload) // 2
+        left = payload[:half]
+        right = payload[half:]
+        salt = b"shamir" + int_to_big_endian(id, 2)
+        for i in (b"\x00", b"\x01", b"\x02", b"\x03"):
+            f = pbkdf2_hmac(
+                "sha256",
+                i + passphrase,
+                salt + right,
+                2500 << exponent,  # 10000 << exponent // 4
```

Comment:
Delete commented out code

Reply:
same as above, though now that I look at it, maybe I can combine them.